src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -4830,46 +4724,63 @@ public long OSThreadId
         {
             get
             {
-                if (!m_osThreadId.HasValue)
+                MoreEventInfo moreInfo = MoreInfo;
+                if (!moreInfo.OsThreadId.HasValue)
                 {
 #if ES_BUILD_STANDALONE
-                    m_osThreadId = (long)Interop.Kernel32.GetCurrentThreadId();
+                    moreInfo.OsThreadId = (long)Interop.Kernel32.GetCurrentThreadId();
 #else
-                    m_osThreadId = (long)Thread.CurrentOSThreadId;
+                    moreInfo.OsThreadId = (long)Thread.CurrentOSThreadId;
 #endif
                 }
 
-                return m_osThreadId.Value;
+                return moreInfo.OsThreadId.Value;
             }
-            internal set => m_osThreadId = value;
+            internal init => MoreInfo.OsThreadId = value;
```

Comment:
`OSThreadId` has a similar issue as `ActivityId` (#42128) where the returned value depends on the caller's thread.

The difference is that `OSThreadId` saves the first caller's value where as `ActivityId` does not. Should these be changed to have the same behavior?

Reply:
Changed